### PR TITLE
Add 'torch' as an image backend and Make VisionDataset respect get_image_backend()

### PR DESCRIPTION
**Pull request by firstprayer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#2607 Add 'torch' as an image backend and Make VisionDataset respect get_image_backend()**

### CODE REVIEW

File: torchvision/datasets/usps.py [hunk 1]
```diff
@@ -75,9 +75,10 @@ def __getitem__(self, index: int) -> Tuple[Any, Any]:
         """
         img, target = self.data[index], int(self.targets[index])
 
-        # doing this so that it is consistent with all other datasets
-        # to return a PIL Image
-        img = Image.fromarray(img, mode='L')
+        if self.get_image_backend() == 'PIL':
+            # doing this so that it is consistent with all other datasets
+            # to return a PIL Image
+            img = Image.fromarray(img, mode='L')
```
Comment (fmassa) [inline torchvision/datasets/usps.py:81]: Does this mean that `img` is in `HWC` format? If that's the case, we would want to make it `CHW` in the `torch` backend, for consistency with how the rest of pytorch (and torchvision now) handles images.

Additionally, isn't `img` a numpy `ndarray`? We would need to convert it to a torch tensor via `torch.as_tensor(img)` in order to have a tensor.

File: torchvision/datasets/semeion.py [hunk 2]
```diff
@@ -60,9 +60,10 @@ def __getitem__(self, index: int) -> Tuple[Any, Any]:
         """
         img, target = self.data[index], int(self.labels[index])
 
-        # doing this so that it is consistent with all other datasets
-        # to return a PIL Image
-        img = Image.fromarray(img, mode='L')
+        if self.get_image_backend() == 'PIL':
+            # doing this so that it is consistent with all other datasets
+            # to return a PIL Image
+            img = Image.fromarray(img, mode='L')
```
Comment (fmassa) [inline torchvision/datasets/semeion.py:66]: Same comment here about the data ordering and ndarray

File: torchvision/datasets/mnist.py [hunk 3]
```diff
@@ -418,7 +419,8 @@ def download(self) -> None:
     def __getitem__(self, index: int) -> Tuple[Any, Any]:
         # redefined to handle the compat flag
         img, target = self.data[index], self.targets[index]
-        img = Image.fromarray(img.numpy(), mode='L')
+        if self.get_image_backend() == 'PIL':
+            img = Image.fromarray(img.numpy(), mode='L')
```
Comment (fmassa) [inline torchvision/datasets/mnist.py:423]: Same comment here about the data ordering and ndarray. Should we be consistent and always return `CHW` format, with C potentially being 1?

File: torchvision/datasets/mnist.py [hunk 4]
```diff
@@ -96,9 +96,10 @@ def __getitem__(self, index: int) -> Tuple[Any, Any]:
         """
         img, target = self.data[index], int(self.targets[index])
 
-        # doing this so that it is consistent with all other datasets
-        # to return a PIL Image
-        img = Image.fromarray(img.numpy(), mode='L')
+        if self.get_image_backend() == 'PIL':
+            # doing this so that it is consistent with all other datasets
+            # to return a PIL Image
+            img = Image.fromarray(img.numpy(), mode='L')
```
Comment (fmassa) [inline torchvision/datasets/mnist.py:102]: Same here

File: torchvision/datasets/cifar.py [hunk 5]
```diff
@@ -112,9 +112,10 @@ def __getitem__(self, index: int) -> Tuple[Any, Any]:
         """
         img, target = self.data[index], self.targets[index]
 
-        # doing this so that it is consistent with all other datasets
-        # to return a PIL Image
-        img = Image.fromarray(img)
+        if self.get_image_backend() == 'PIL':
+            # doing this so that it is consistent with all other datasets
+            # to return a PIL Image
+            img = Image.fromarray(img)
```
Comment (fmassa) [inline torchvision/datasets/cifar.py:118]: Same comment here

File: torchvision/datasets/cifar.py [hunk 6]
```diff
@@ -112,9 +112,10 @@ def __getitem__(self, index: int) -> Tuple[Any, Any]:
         """
         img, target = self.data[index], self.targets[index]
 
-        # doing this so that it is consistent with all other datasets
-        # to return a PIL Image
-        img = Image.fromarray(img)
+        if self.get_image_backend() == 'PIL':
```
Comment (fmassa) [inline torchvision/datasets/cifar.py:115]: This assumes that we only have two backends, right? What if the user had specified `accimage`?

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

I have left a few comments, I think there are things we could improve.

Also, we might want to add a `torch_loader` implementation for our new `io` reading ops, which are used in `ImageFolder` for example https://github.com/pytorch/vision/blob/fc69c22576cbccb59c581ddbaca4dedbdb279688/torchvision/datasets/folder.py#L174-L188

File: torchvision/datasets/usps.py (hunk 1 above)
Comment (firstprayer) [reply torchvision/datasets/usps.py:81]: MNIST is already Tensor in itself I think, I haven't checked the other datasets. Will check them all.

**Pull request closed without merging**